### PR TITLE
Add NumMemoryStateInputsOutputs annotation to RvsdgTreePrinter

### PR DESCRIPTION
**Pull request by phate**
1. Extends the structural test node with an interface for adding inputs/outputs/arguments/results
2. Add support for printing of memory state typed inputs/outputs to RvsdgTreePrinter class
3. Adds support for this annotation to jlm-opt

### CODE REVIEW

File: jlm/llvm/opt/RvsdgTreePrinter.cpp [hunk 1]
```diff
@@ -115,6 +118,78 @@ RvsdgTreePrinter::AnnotateNumRvsdgNodes(
...
+        annotationMap.AddAnnotation(structuralNode, { outputLabel, numMemoryStateOutputs });
+
+        for (size_t n = 0; n < structuralNode->nsubregions(); n++)
+        {
+          auto subregion = structuralNode->subregion(n);
+          annotateRegion(*subregion);
+        }
+      }
+    }
+  };
```
Comment (phate) [inline jlm/llvm/opt/RvsdgTreePrinter.cpp:188]: There is a lot of repetition in here. However, in order to remove this, it would be necessary to add better support for iterators/iterator ranges to the base classes.

File: tests/test-operation.hpp [hunk 2]
```diff
@@ -205,6 +205,10 @@ class structural_op final : public rvsdg::structural_op
   copy() const override;
 };
```
Comment (phate) [inline tests/test-operation.hpp:207]: The entire extension to the test node could be put in its own PR if desired.

File: jlm/llvm/opt/RvsdgTreePrinter.cpp [hunk 3]
```diff
@@ -115,6 +118,78 @@ RvsdgTreePrinter::AnnotateNumRvsdgNodes(
...
+RvsdgTreePrinter::AnnotateNumMemoryStateInputsOutputs(
+    const rvsdg::graph & rvsdg,
+    util::AnnotationMap & annotationMap)
+{
+  static std::string_view argumentLabel("NumMemoryStateTypeArguments");
+  static std::string_view resultLabel("NumMemoryStateTypeResults");
+  static std::string_view inputLabel("NumMemoryStateTypeInputs");
+  static std::string_view outputLabel("NumMemoryStateTypeOutputs");
+
+  std::function<void(const rvsdg::region &)> annotateRegion = [&](const rvsdg::region & region)
```
Comment (haved) [inline jlm/llvm/opt/RvsdgTreePrinter.cpp:131]: Is there a reason the lambda is stored as a function pointer in `std::function`, and not an `auto` functor?

Comment (haved) [reply jlm/llvm/opt/RvsdgTreePrinter.cpp:131]: Ah, I see now it is recursive. Nice trick, I didn't know you could do recursive lambdas like that

File: jlm/llvm/opt/RvsdgTreePrinter.cpp [hunk 4]
```diff
@@ -115,6 +118,78 @@ RvsdgTreePrinter::AnnotateNumRvsdgNodes(
...
 
+void
+RvsdgTreePrinter::AnnotateNumMemoryStateInputsOutputs(
+    const rvsdg::graph & rvsdg,
+    util::AnnotationMap & annotationMap)
+{
+  static std::string_view argumentLabel("NumMemoryStateTypeArguments");
+  static std::string_view resultLabel("NumMemoryStateTypeResults");
+  static std::string_view inputLabel("NumMemoryStateTypeInputs");
+  static std::string_view outputLabel("NumMemoryStateTypeOutputs");
```
Comment (haved) [inline jlm/llvm/opt/RvsdgTreePrinter.cpp:129]: I checked compiler explorer, and it is better to leave these as non-static. It does not do a strlen at runtime either way, but making them static adds an extra level of indirection to access the length and string data.

**Review by haved — APPROVED** [review]
Looks good. My only comment is that we can save 16 bytes and some loads in the final binary by not making the `string_view`s static.

File: jlm/llvm/opt/RvsdgTreePrinter.cpp (hunk 3 above)
Comment (phate) [reply jlm/llvm/opt/RvsdgTreePrinter.cpp:131]: You answered it yourself. :)

File: jlm/llvm/opt/RvsdgTreePrinter.cpp (hunk 4 above)
Comment (phate) [reply jlm/llvm/opt/RvsdgTreePrinter.cpp]: I removed that `static`s. Could you approve again, please?

**Review by haved — APPROVED** [review]
